### PR TITLE
Support of Embedded and XXXToOne nested object projection + Support of @ProjectedFieldName on class fields

### LINKED ISSUE #36581: Support of Embedded and ManyToOne nested object projection + Support of @ProjectedFieldName on class fields
### Description

1) Add support of @Embedded and @XXXToOne nested objects in panache query projection.

2) When you use Lombok for DTO, it could be helpfull to have the @ProjectedFieldName annotation available for fields instead of constructor parameters.

### Implementation ideas

Add support in CommonPanacheQueryImpl#project

### PR DESCRIPTION
**Pull request by humcqc**
Fixes #36581 

Support of Embedded and ManyToOne nested object projection + Support of @ProjectedFieldName on class fields

Documentation not updated yet, will do it when we agree on the code.

### CODE REVIEW

**Comment by loicmathieu** [comment]
If I understand it correctly, the `@ProjectedClass` annotation allows for nested objects into a projection class. Can you provide a simple example?

Also, the documentation for projection should be updated.

**Comment by humcqc** [comment]
https://github.com/quarkusio/quarkus/pull/36594/files#diff-41e4522791d7b2934d3189b9d2608d935564a31102f309bfb7f22e27e7aed968R56

**Comment by loicmathieu** [comment]
What I don't understand in your example is why the PersonDTO extends PeronsName and why it has the annotation in it. That's why I asked for a simple example.

**Comment by humcqc** [comment]
> What I don't understand in your example is why the PersonDTO extends PeronsName and why it has the annotation in it. That's why I asked for a simple example.

PersonName is the original DTO in integration test , I ve created PersonDTO that extend it to keep the original test and I've  included Embbeded (DescriptionDTO) and ManyToOne ( AdressDTO) elements in it. It allow me to test different case with this DTO. 

"Why there is an annotation in it ?" Good question. We have 3 choices:
1) Impose the @ProjectedClass for all projected class and test if it present ( breaking change )
2) Does not impose it, you can put it but the ProjectedClass annotation on main class have no impact -> what we have in the integration test.
3) Just use it in nested class , and I can remove it from Main class

**Comment by loicmathieu** [comment]
> Just use it in nested class , and I can remove it from Main class

It is clearer to me, its behavior will be well defined.

> PersonName is the original DTO in integration test , I ve created PersonDTO that extend it to keep the original test and I've included Embbeded (DescriptionDTO) and ManyToOne ( AdressDTO) elements in it. It allow me to test different case with this DTO.

Please let the current test as is and add a new one so we test both a simple projection and a more complex one.

**Comment by humcqc** [comment]
> > Just use it in nested class , and I can remove it from Main class
> 
> It is clearer to me, its behavior will be well defined.

---> Ok , perhaps we can rename it to NestedProjectedClass to be more precise ?

> > PersonName is the original DTO in integration test , I ve created PersonDTO that extend it to keep the original test and I've included Embbeded (DescriptionDTO) and ManyToOne ( AdressDTO) elements in it. It allow me to test different case with this DTO.
> 
> Please let the current test as is and add a new one so we test both a simple projection and a more complex one.

--> I didn't change the current test with simple Projection ( PersonName), it's just my test use a new DTO ( PersonDTO) that inherit from the simple (PersonName) to avoid to duplicate the projection of simple field. And to test the inheritance, I had some issue with inheritance with the constructor part. 
If you think it's simpler I can create a new class PersonName2 same as PersonName and PersonDTO will inherit this new Class ?

**Comment by humcqc** [comment]
I've added simple example + doc, and renamed annotation.

**Comment by humcqc** [comment]
Let me fix the jdk11 + native issue and i will come back to you.

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/Person.java [hunk 1]
```diff
@@ -46,8 +36,11 @@
 })
 @FilterDef(name = "Person.hasName", defaultCondition = "name = :name", parameters = @ParamDef(name = "name", type = String.class))
 @FilterDef(name = "Person.isAlive", defaultCondition = "status = 'LIVING'")
+@FilterDef(name = "Person.name.in", defaultCondition = "name in (:names)", parameters = {
+        @ParamDef(name = "names", type = String.class) })
```
Comment (loicmathieu) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/Person.java:40]: Why did you add those filter?

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/Person.java [hunk 2]
```diff
@@ -92,4 +88,8 @@ public void setSerialisationTrick(int serialisationTrick) {
     public static long methodWithPrimitiveParams(boolean b, byte bb, short s, int i, long l, float f, double d, char c) {
         return 0;
     }
+
+    public static void voidMethod() {
+        throw new RuntimeException("void");
+    }
```
Comment (loicmathieu) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/Person.java:94]: What is the purpose of this method?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Overall, the PR seems more complex than needed.
There seem to be multiple functionalities in one PR and the tests seem contain more than just what's needed for the functionality.

This makes it complex to review so I didn't finish it yet.

**Comment by humcqc** [comment]
> Overall, the PR seems more complex than needed. There seem to be multiple functionalities in one PR and the tests seem contain more than just what's needed for the functionality.
> 
> This makes it complex to review so I didn't finish it yet.

Yes you right , the PR is complex because it handles the recursivity of nested objects. And I included one test for #28844.
I will simplify and remove the test for #28844.

**Comment by humcqc** [comment]
Hi @loicmathieu , It should be good now, could you have a look please?

File: extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java [hunk 3]
```diff
@@ -121,39 +121,73 @@ public <T> CommonPanacheQueryImpl<T> project(Class<T> type) {
             return new CommonPanacheQueryImpl<>(this, newQuery.toString(), "select count(*) " + from);
         }
 
+        // build select clause with a constructor expression
+        String selectClause = "SELECT " + getParameterFromClass(type, null);
+        return new CommonPanacheQueryImpl<>(this, selectClause + selectQuery,
+                "select count(*) " + selectQuery);
+    }
+
+    private StringBuilder getParameterFromClass(Class<?> type, String parentParameter) {
```
Comment (loicmathieu) [inline extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:130]:
```suggestion
    private StringBuilder getParametersFromClass(Class<?> type, String parentParameter) {
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by humcqc** [comment]
Hi @loicmathieu , could you check please to finalize this PR ? Thanks

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DogDto2.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    public String name;
+    public PersonDto2 owner;
+
+    public DogDto2(String name, PersonDto2 owner) {
+        this.name = name;
+        this.owner = owner;
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/DogDto2.java:17]: Is it really needed?
If yes we should register for reflection all classes annotated by `@NestedProjectedClass`

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/PersonDTO.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    public PersonDTO(String uniqueName, String name, AddressDTO address, DescriptionDTO description,
+            Integer directHeight) {
+        super(uniqueName, name);
+        this.address = address;
+        this.description = description;
+        this.directHeight = directHeight;
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/PersonDTO.java:26]: Same here

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/PersonDTO.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+        @Override
+        public int compareTo(AddressDTO address) {
+            return street.compareTo(address.street);
+        }
+
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/PersonDTO.java:45]: Same here

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/DogDto.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    public String name;
+    public PersonDto owner;
+
+    public DogDto(String name, PersonDto owner) {
+        this.name = name;
+        this.owner = owner;
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/DogDto.java:17]: Same here

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/PersonDTO.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    public PersonDTO(String uniqueName, String name, AddressDTO address, DescriptionDTO description,
+            Integer directHeight) {
+        super(uniqueName, name);
+        this.address = address;
+        this.description = description;
+        this.directHeight = directHeight;
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/PersonDTO.java:25]: Same here

File: integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/PersonDTO.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+        @Override
+        public int compareTo(AddressDTO address) {
+            return street.compareTo(address.street);
+        }
+
+    }
+
+    @NestedProjectedClass
+    @RegisterForReflection
```
Comment (loicmathieu) [inline integration-tests/hibernate-reactive-panache/src/main/java/io/quarkus/it/panache/reactive/PersonDTO.java:42]: Same here

File: integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheFunctionalityTest.java [hunk 10]
```diff
@@ -235,7 +236,118 @@ Person getBug7102(Long id) {
...
+        person.name = "2";
+        person.uniqueName = "2";
+        person.address = new Address("street 2");
+        person.persist();
+        PersonDTO personDTO = Person.find(
+                "select uniqueName, name, " +
+                        " new io.quarkus.it.panache.PersonDTO$AddressDTO(address.street)," +
+                        " new io.quarkus.it.panache.PersonDTO$DescriptionDTO(description.size, description.weight)," +
+                        " description.size" +
+                        " from Person2 where name = ?1",
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheFunctionalityTest.java:297]: I'm surprise by this, the project method should built the select clause so why adding it manually here?

File: integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java [hunk 11]
```diff
@@ -1355,6 +1355,64 @@ public String testProjection() {
...
+        person.name = "2n";
+        person.uniqueName = "2n";
+        person.address = new Address("street 2");
+        person.persist();
+        PersonDTO personDTO = Person.find(
+                "select uniqueName, name, " +
+                        " new io.quarkus.it.panache.PersonDTO$AddressDTO(address.street)," +
+                        " new io.quarkus.it.panache.PersonDTO$DescriptionDTO(description.size, description.weight)," +
+                        " description.size" +
+                        " from Person2 where name = ?1",
```
Comment (loicmathieu) [inline integration-tests/hibernate-orm-panache/src/main/java/io/quarkus/it/panache/TestEndpoint.java:1372]: The select clause should be build by the project method so why add one here?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
Two points still to solve:
- A lot of tests are disabled in native, why didn't they work and is there a way to make them working?
- If a class is annotated by our own annotation, we can easily register them for reflection so we should not ask our user to do that. If it's needed to register them, you can do it in the processor, I can help for that.

**Comment by humcqc** [comment]
For the native test i've followed the same pattern in other test in PanacheFunctionalityTest : https://github.com/quarkusio/quarkus/blob/76a7ce15569b56fdacf88a4a4d42fd02e4bf551a/integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/PanacheFunctionalityTest.java#L108 , seems just Rest Tests are working in this test class. I can remove them from PanacheFunctionalityTest as they are tested through TestEnpoint in native mode ?

**Comment by loicmathieu** [comment]
> Seems just Rest Tests are working in native mode in this test class.I would rather keep them , but if you really want i can remove them from PanacheFunctionalityTest as they are tested through TestEnpoint in native mode ?

Yes please

**Comment by humcqc** [comment]
If a class is annotated by our own annotation, we can easily register them for reflection so we should not ask our user to do that. If it's needed to register them, you can do it in the processor, I can help for that.

I would appreciate some help for this one. I try a simple fix, but did not work :)

**Comment by humcqc** [comment]
Hi @loicmathieu , could we try to finish this week to have it in 3.7 ?

**Comment by humcqc** [comment]
Hi @loicmathieu @geoand ,
I think this PR will be fine when [38496](https://github.com/quarkusio/quarkus/pull/38496#issuecomment-1918938699) will ne done.
Could we try to have it in 3.7 ? 
Thanks

**Comment by geoand** [comment]
Thanks for this. 

We will certainly need to the commits to be cleaned up / squashed

**Review by loicmathieu — APPROVED** [review]
LGTM, thanks for doing this and your patience for the long review process!

**Comment by loicmathieu** [comment]
@humcqc can you please squash your commits in one commit so we can merge it?

**Comment by humcqc** [comment]
> @humcqc can you please squash your commits in one commit so we can merge it?



> @humcqc can you please squash your commits in one commit so we can merge it?

Sure, Done!

**Pull request merged**